### PR TITLE
feat: compile time middleware

### PR DESCRIPTION
**Pull request by crookse**
Addresses #413 

## Description

* Adds logic to execute compile time middleware. When compile time middleware is executed, runtime middleware gets set to run at a later time (only after requests -- i don't see a reason from runtime data to be executed before a request).
* https://github.com/drashland/deno-drash-middleware/issues/59 is dependent on this feature.

### CODE REVIEW

File: src/http/server.ts [hunk 1]
```diff
@@ -471,9 +482,26 @@ export class Server {
       // requested asset. Since this occurs after the MIME code above, the
       // client should receive a proper response in the proper format.
       if (!this.configs.pretty_links || request.url.split(".")[1]) {
-        response.body = Deno.readFileSync(`${this.directory}/${request.url}`);
-        await this.executeMiddlewareServerLevelAfterRequest(request, response);
-        return response.sendStatic();
+        try {
+          // Try to reaed the file if it exists
```
Comment (ebebbington) [inline src/http/server.ts:486]: *reaed*

File: src/http/server.ts [hunk 2]
```diff
@@ -851,15 +892,56 @@ export class Server {
...
+   * @param response - The response object.
+   *
+   * @returns A Drash response or a boolean. If it returns a boolean, it means
+   * it did or did not process properly (true being it processed; false being it
+   * did not process).
+   */
+  protected async executeMiddlewareServerLevelRuntimeAfterRequest(
+    request: Drash.Http.Request,
+    response: Drash.Http.Response,
+  ): Promise<void> {
```
Comment (ebebbington) [inline src/http/server.ts:927]: Doc block says it returns a drash response or boolean, but method return type is void

File: src/http/server.ts [hunk 3]
```diff
@@ -303,6 +313,7 @@ export class Server {
         throw new Drash.Exceptions.HttpResponseException(
           418,
           "The response must be returned inside the " +
+            resource.constructor.name + "." +
             request.method.toUpperCase() + " method of the resource",
```
Comment (ebebbington) [inline src/http/server.ts:317]: This results in `The response must be returned inside the HomeResource.GET method of the resource`, wonder if it sounds better to you for it to be `The response must be returned inside the GET method of the "HomeResource"`?

**Review by ebebbington — CHANGES_REQUESTED** [review]
Code looks good to me, a few comments, but i'm still unsure how a user would go about using ts on the client without transpiling and using compile time middleware?

File: src/http/server.ts (hunk 3 above)
Comment (crookse) [reply src/http/server.ts:317]: that makes more sense

Comment (crookse) [reply src/http/server.ts]: 0aa19c1

File: src/http/server.ts (hunk 2 above)
Comment (crookse) [reply src/http/server.ts]: 93082a1 (it supposed to return void -- i removed the @return part)

File: src/http/server.ts (hunk 1 above)
Comment (crookse) [reply src/http/server.ts]: 7c1d0c8 (ok it READs better now)

**Review by ebebbington — APPROVED** [review]
image

**Pull request merged**